### ISSUE TITLE
Consider adding meta.schema

### ISSUE THREAD

**Issue by TKDev7** [stylelint/stylelint#7608]
It would be equivalent to [ESLint’s meta.schema](https://eslint.org/docs/latest/extend/custom-rules#options-schemas). This can be used for generating plugin/rule documentation websites and typegens.

**Comment by ybiquitous** [comment]
Sounds good 👍🏼 

Maybe one downside is increasing a dependency that processes the JSON schema, right?

**Comment by ybiquitous** [comment]
Great. `ajv` is already included in our runtime dependency tree (`stylelint` depends on `table`).

```sh-session
$ npm ls ajv
stylelint@16.3.1 /Users/masafumi.koba/git/stylelint/stylelint
├─┬ eslint@8.57.0
│ ├─┬ @eslint/eslintrc@2.1.4
│ │ └── ajv@6.12.6 deduped
│ └── ajv@6.12.6
└─┬ table@6.8.2
  └── ajv@8.12.0
```

**Comment by ybiquitous** [comment]
@MrHBS If you're interested, could you please show more specific idea to implement this feature so that others could image it easily?

**Comment by ybiquitous** [comment]
@MrHBS Thanks for the helpful info.

@ota-meshi Do you have any ideas to bring this feature from ESLint?

**Comment by Mouvedia** [comment]
Here's another use case for autocompletion: [knowing the message arguments' names and positions](https://github.com/stylelint/stylelint/pull/6312/files#r968253349).

**Comment by Mouvedia** [comment]
I am working on a pre-requisite for `schema` that will expedite the conversion to JSON schema.

**Comment by Mouvedia** [comment]
@MrHBS

I have a few questions.

1. Would `schema` be required if the rule is not `CoreRule<true>`?
i.e. doesn't accept `true` or `null` as a primary and global secondary options

> [Rules with options must specify a meta.schema property…](https://eslint.org/docs/latest/extend/custom-rules#options-schemas)

2. If introduced, would it sunset the `...optionDescriptions` argument of the `validateOptions` function?
i.e. `possible` and `optional` would be covered by the schema

3. Would introducing a way to generate a schema for the current config be sufficient?
e.g.
- `stylelint schema > foo.jtd.json` defaults to JSON Type Definition (default `--type` is `"RFC8927"`)
- `stylelint schema -c ../bar/.stylelintrc > foo.jtd.json` doesn't use the default config
- `stylelint schema --type="2020-12" > foo.schema.json` sets the format to JSON Schema version 2020-12

---

I am asking all that because I am under the impression that deprecating the current `validateOptions` util would be too disruptive.
i.e.
Just providing `stylelint schema` should be enough for the tools—e.g. `stylelint-doc-generator`, `stylelint-typegen`—to leverage the schema. Adding `meta.schema` would have cascading consequences:
- dropping `validateOptions` (which does its job just fine)
- making `meta.schema` a requirement for most rules (would have to wait the next major)
- choosing a default format (less versatile than slowly introducing new types to the `--type` option)

tl;dr: I think it can be achieved _easily_ but through a different mean.
